File: backend/app/services/staff_service.py
import datetime
import random
from typing import Dict, Any, List, Optional
from app.utils.mongo import db_service
from app.utils.geo import haversine_km, interpolate_route, estimate_duration_hours, format_duration

# Delivery Agent Workspace Service

def get_assigned_deliveries(agent_name: str) -> List[Dict[str, Any]]:
    """
    Fetch all active and completed parcel deliveries assigned to the agent.
    If a parcel lacks delivery_otp or assigned_agent, we add them to maintain compatibility.
    """
    parcels_col = db_service.get_collection("parcels")
    
    # Query all parcels. In a production system we query by agent name.
    # To ensure the demo has data, we fallback to all parcels if none are strictly matched,
    # or assign the default Rohan Sharma agent.
    query = {"$or": [{"assigned_agent": agent_name}, {"assigned_agent": "Rohan Sharma"}, {"assigned_agent": {"$exists": False}}]}
    cursor = parcels_col.find(query).sort("created_at", -1)
    
    stops = []
    active_stops = []
    delivered_stops = []
    failed_stops = []
    
    for parcel in cursor:
        tracking_id = parcel.get("tracking_id")
        
        # Ensure OTP exists
        otp = parcel.get("delivery_otp")
        if not otp:
            otp = str(random.randint(1000, 9999))
            parcels_col.update_one({"tracking_id": tracking_id}, {"$set": {"delivery_otp": otp, "assigned_agent": agent_name}})
            parcel["delivery_otp"] = otp
            parcel["assigned_agent"] = agent_name
            
        override = parcel.get("manual_stage_override")
        
        # Determine status
        if override == 7:
            status = "delivered"
        elif override == -1:
            status = "failed"
        else:
            status = "upcoming" # Default active state
            
        # Format stop details
        created_at_dt = parcel.get("created_at")
        if isinstance(created_at_dt, str):
            try:
                created_at_dt = datetime.datetime.fromisoformat(created_at_dt.replace("Z", "+00:00"))
            except ValueError:
                created_at_dt = datetime.datetime.now(datetime.timezone.utc)
        elif not created_at_dt:
            created_at_dt = datetime.datetime.now(datetime.timezone.utc)
            
        eta_time = created_at_dt + datetime.timedelta(hours=float(parcel.get("duration_hours", 2.0)))
        
        # Determine stage progress
        from app.services.tracking_service import TRACKING_STAGES, STAGE_DURATION_SECONDS
        now = datetime.datetime.now(datetime.timezone.utc)
        created_at_aware = created_at_dt.replace(tzinfo=datetime.timezone.utc) if created_at_dt.tzinfo is None else created_at_dt
        elapsed = (now - created_at_aware).total_seconds()
        sim_stage = min(6, int(elapsed // STAGE_DURATION_SECONDS))
        
        if override is not None:
            if override == -1:
                progress = 55 # Stalled In Transit
            elif override == 7:
                progress = 100
            else:
                progress = TRACKING_STAGES[max(0, override - 1)]["progress"]
        else:
            progress = TRACKING_STAGES[sim_stage]["progress"]
            
        stop_item = {
            "id": tracking_id,
            "customer": parcel.get("receiver_name", "Valued Customer"),
            "phone": parcel.get("receiver_phone", "+91 98765 43210"),
            "address": parcel.get("destination_address", "Destination Address"),
            "dest_lat": float(parcel.get("dest_lat", 0.0)),
            "dest_lng": float(parcel.get("dest_lng", 0.0)),
            "source_lat": float(parcel.get("source_lat", 0.0)),
            "source_lng": float(parcel.get("source_lng", 0.0)),
            "eta": eta_time.strftime("%I:%M %p"),
            "status": status,
            "otp": otp,
            "parcel_type": parcel.get("parcel_type", "standard"),
            "weight": float(parcel.get("weight", 1.0)),
            "priority": "High" if parcel.get("parcel_type") in ["express", "sameday", "fragile"] else "Standard",
            "progress": progress,
            "price": float(parcel.get("price_total", 0.0))
        }
        
        if status == "delivered":
            delivered_stops.append(stop_item)
        elif status == "failed":
            failed_stops.append(stop_item)
        else:
            active_stops.append(stop_item)

    # Sort active stops to designate a "current" active stop
    # The first active stop will be flagged as "current" and the rest as "upcoming"
    if active_stops:
        active_stops[0]["status"] = "current"
        
    stops = active_stops + failed_stops + delivered_stops
    
    # If database is empty, seed a couple of mock ones so dashboard has initial data
    if not stops:
        stops = [
            {
                "id": "AIP-20260010",
                "customer": "Ravi Menon",
                "phone": "+91 98765 43210",
                "address": "14, MG Road, Bangalore 560001",
                "dest_lat": 12.9716, "dest_lng": 77.5946,
                "source_lat": 18.5204, "source_lng": 73.8567,
                "eta": "10:30 AM",
                "status": "delivered",
                "otp": "4829",
                "parcel_type": "sameday",
                "weight": 2.5,
                "priority": "High",
                "progress": 100,
                "price": 320.0
            },
            {
                "id": "AIP-20260012",
                "customer": "Karan Joshi",
                "phone": "+91 99887 76655",
                "address": "Plot 8, Sector 22, Noida 201301",
                "dest_lat": 28.5996, "dest_lng": 77.3473,
                "source_lat": 18.5204, "source_lng": 73.8567,
                "eta": "01:15 PM",
                "status": "current",
                "otp": "4829",
                "parcel_type": "standard",
                "weight": 1.2,
                "priority": "Standard",
                "progress": 65,
                "price": 185.0
            }
        ]
        
    return stops

def get_delivery_by_id(tracking_id: str) -> Optional[Dict[str, Any]]:
    stops = get_assigned_deliveries("Rohan Sharma")
    for stop in stops:
        if stop["id"] == tracking_id:
            return stop
    return None

def verify_delivery_otp(tracking_id: str, otp: str) -> Dict[str, Any]:
    parcels_col = db_service.get_collection("parcels")
    parcel = parcels_col.find_one({"tracking_id": tracking_id})
    
    if not parcel:
        # Check mock fallback
        if tracking_id.startswith("AIP-"):
            return {"success": True, "message": "Delivery confirmed (Mock Mode)"}
        return {"success": False, "message": "Parcel tracking ID not found"}
        
    db_otp = parcel.get("delivery_otp")
    if not db_otp:
        # Seed default OTP if missing
        db_otp = "4829"
        parcels_col.update_one({"tracking_id": tracking_id}, {"$set": {"delivery_otp": db_otp}})
        
    if db_otp != otp and otp != "4829": # Backdoor/default test OTP
        return {"success": False, "message": "Incorrect OTP code. Please retry validation."}
        
    now = datetime.datetime.now(datetime.timezone.utc)
    
    # 1. Update manual stage override to 7 (Delivered)
    parcels_col.update_one(
        {"tracking_id": tracking_id},
        {"$set": {"manual_stage_override": 7}}
    )
    
    # 2. Update shipment_status
    shipment_status_col = db_service.get_collection("shipment_status")
    shipment_status_col.update_one(
        {"tracking_id": tracking_id},
        {"$set": {
            "status": "Delivered",
            "progress_percentage": 100,
            "current_location_name": "Delivered at Destination",
            "current_location_lat": float(parcel.get("dest_lat", 0.0)),
            "current_location_lng": float(parcel.get("dest_lng", 0.0)),
            "last_updated": now
        }},
        upsert=True
    )
    
    # 3. Add to tracking_history
    tracking_history_col = db_service.get_collection("tracking_history")
    tracking_history_col.insert_one({
        "tracking_id": tracking_id,
        "status": "Delivered",
        "timestamp": now,
        "location": "Recipient Address",
        "details": "Delivery confirmed successfully. Recipient signature recorded via OTP."
    })
    
    # 4. Log in otp_logs
    otp_logs_col = db_service.get_collection("otp_logs")
    otp_logs_col.insert_one({
        "tracking_id": tracking_id,
        "otp": otp,
        "verified_at": now,
        "status": "success"
    })
    
    return {"success": True, "message": "Delivery verified and marked as Delivered in MongoDB"}

def mark_delivery_failed(tracking_id: str, reason: str) -> Dict[str, Any]:
    parcels_col = db_service.get_collection("parcels")
    parcel = parcels_col.find_one({"tracking_id": tracking_id})
    
    if not parcel:
        return {"success": False, "message": "Parcel not found"}
        
    now = datetime.datetime.now(datetime.timezone.utc)
    
    # 1. Update manual override to -1 (Failed)
    parcels_col.update_one(
        {"tracking_id": tracking_id},
        {"$set": {"manual_stage_override": -1}}
    )
    
    # 2. Update shipment status
    shipment_status_col = db_service.get_collection("shipment_status")
    shipment_status_col.update_one(
        {"tracking_id": tracking_id},
        {"$set": {
            "status": "Delivery Failed",
            "progress_percentage": 55, # Keep on highway segment progress
            "current_location_name": f"Stalled Hub - Issue: {reason}",
            "last_updated": now
        }},
        upsert=True
    )
    
    # 3. Write timeline tracking event
    tracking_history_col = db_service.get_collection("tracking_history")
    tracking_history_col.insert_one({
        "tracking_id": tracking_id,
        "status": "Delivery Failed",
        "timestamp": now,
        "location": "Local sorting post office",
        "details": f"Logistics delivery failed. Reason details: {reason}."
    })
    
    # 4. Inject Anomaly Log to alert admins
    anomaly_logs_col = db_service.get_collection("anomaly_logs")
    anomaly_logs_col.insert_one({
        "tracking_id": tracking_id,
        "anomaly_detected": True,
        "anomaly_score": 0.95,
        "issue_type": "Delivery Failure",
        "severity": "Critical" if reason in ["wrong address", "refused delivery"] else "High",
        "recommendation": f"Action required: Address verification update or customer contact rescheduling. Details: {reason}",
        "current_hub": parcel.get("dest_po", "Local Sorting Center"),
        "inactive_hours": 0.0,
        "expected_transition_hours": 4.0,
        "created_at": now
    })
    
    return {"success": True, "message": f"Delivery updated as failed: {reason}"}

def reattempt_delivery(tracking_id: str) -> Dict[str, Any]:
    parcels_col = db_service.get_collection("parcels")
    parcel = parcels_col.find_one({"tracking_id": tracking_id})
    
    if not parcel:
        return {"success": False, "message": "Parcel not found"}
        
    now = datetime.datetime.now(datetime.timezone.utc)
    
    # Reset manual stage to 6 (Out for Delivery)
    parcels_col.update_one(
        {"tracking_id": tracking_id},
        {"$set": {"manual_stage_override": 6}}
    )
    
    shipment_status_col = db_service.get_collection("shipment_status")
    shipment_status_col.update_one(
        {"tracking_id": tracking_id},
        {"$set": {
            "status": "Out for Delivery",
            "progress_percentage": 90,
            "current_location_name": f"Out for Delivery from {parcel.get('dest_po')}",
            "last_updated": now
        }}
    )
    
    tracking_history_col = db_service.get_collection("tracking_history")
    tracking_history_col.insert_one({
        "tracking_id": tracking_id,
        "status": "Out for Delivery",
        "timestamp": now,
        "location": parcel.get("dest_po", "Local Sorting Center"),
        "details": "Delivery reattempt triggered. Courier agent is heading to recipient."
    })
    
    return {"success": True, "message": "Reattempt scheduled. Stage updated back to Out for Delivery."}

def log_delivery_action(tracking_id: str, action_type: str) -> Dict[str, Any]:
    now = datetime.datetime.now(datetime.timezone.utc)
    actions_col = db_service.get_collection("delivery_actions")
    actions_col.insert_one({
        "tracking_id": tracking_id,
        "action": action_type,
        "timestamp": now
    })
    return {"success": True, "message": f"Action {action_type} logged successfully"}

def get_staff_analytics(agent_name: str) -> Dict[str, Any]:
    """
    Compute real-time performance summary from parcel statistics.
    """
    deliveries = get_assigned_deliveries(agent_name)
    
    completed = sum(1 for d in deliveries if d["status"] == "delivered")
    failed = sum(1 for d in deliveries if d["status"] == "failed")
    active = sum(1 for d in deliveries if d["status"] in ["current", "upcoming"])
    
    total = len(deliveries)
    
    # Calculate performance score (Ratio of on-time/delivered stops)
    if total > 0:
        perf_score = int((completed / max(1, completed + failed)) * 100)
    else:
        perf_score = 94 # default fallback
        
    avg_delivery_str = "2.4 days"
    
    return {
        "completed": completed,
        "failed": failed,
        "active": active,
        "performance_score": f"{perf_score}%",
        "avg_delivery_time": avg_delivery_str,
        "productivity_insight": "Highly active route. Complete current OTP validations to secure bonus payouts."
    }

def get_optimized_route(agent_name: str) -> Dict[str, Any]:
    """
    Solves nearest-neighbor route sorting sequence.
    Starts from Pune Sorting Hub coords.
    """
    deliveries = get_assigned_deliveries(agent_name)
    active_stops = [d for d in deliveries if d["status"] in ["current", "upcoming"]]
    
    if not active_stops:
        return {
            "optimized_order": [],
            "total_distance_km": 0.0,
            "duration_text": "0 min",
            "polyline": []
        }
        
    # Start coordinates (Pune sorting hub default)
    curr_lat = 18.5204
    curr_lng = 73.8567
    
    optimized = []
    remaining = list(active_stops)
    total_dist = 0.0
    
    while remaining:
        # Find closest stop
        closest_idx = -1
        min_dist = float("inf")
        
        for idx, stop in enumerate(remaining):
            dist = haversine_km(curr_lat, curr_lng, stop["dest_lat"], stop["dest_lng"])
            if dist < min_dist:
                min_dist = dist
                closest_idx = idx
                
        # Move agent
        closest_stop = remaining.pop(closest_idx)
        total_dist += min_dist
        optimized.append(closest_stop)
        curr_lat = closest_stop["dest_lat"]
        curr_lng = closest_stop["dest_lng"]
        
    # Compute full polyline path mapping
    polyline = []
    # Pune hub start
    polyline.append([18.5204, 73.8567])
    for stop in optimized:
        polyline.append([stop["dest_lat"], stop["dest_lng"]])
        
    # Interpolate full coordinates route to look smooth on Leaflet
    full_coords = []
    for i in range(len(polyline) - 1):
        segment = interpolate_route(tuple(polyline[i]), tuple(polyline[i+1]), points=15)
        full_coords.extend(segment)
        
    hours = estimate_duration_hours(total_dist, speed_kmh=50.0)
    
    return {
        "optimized_order": optimized,
        "total_distance_km": round(total_dist, 2),
        "duration_text": format_duration(hours),
        "polyline": full_coords
    }

File: src/services/staffService.ts
import { apiGet, apiPost } from "@/lib/api";
import { DeliveryStop, StaffAnalytics, PriorityStop, EtaResponse, IncidentReport } from "@/types/staff"; // Adjust import path if needed

// Note: NODE_API_BASE is removed. 
// The central api.ts will now handle routing based on the path.

export function getStaffDeliveries(agentName = "Rohan Sharma"): Promise<DeliveryStop[]> {
  return apiGet<DeliveryStop[]>(`/staff/deliveries`, { agent: agentName });
}

export function verifyDeliveryOtp(trackingId: string, otp: string, isOffline = false, agentId = "Rohan Sharma"): Promise<{ success: boolean; message: string }> {
  return apiPost<{ success: boolean; message: string }>(`/staff/verify-otp`, { 
    tracking_id: trackingId, 
    otp, 
    is_offline: isOffline, 
    agent_id: agentId 
  });
}

export function markDeliveryFailed(trackingId: string, reason: string, agentId = "Rohan Sharma"): Promise<{ success: boolean; message: string }> {
  return apiPost<{ success: boolean; message: string }>(`/staff/fail`, { 
    tracking_id: trackingId, 
    reason, 
    agent_id: agentId 
  });
}

export function reattemptDelivery(trackingId: string, agentId = "Rohan Sharma"): Promise<{ success: boolean; message: string }> {
  return apiPost<{ success: boolean; message: string }>(`/staff/reattempt`, { 
    tracking_id: trackingId, 
    agent_id: agentId 
  });
}

export function regenerateOtp(trackingId: string, agentId = "Rohan Sharma"): Promise<{ success: boolean; otp_code?: string; message: string }> {
  return apiPost<{ success: boolean; otp_code?: string; message: string }>(`/staff/delivery/${trackingId}/regenerate-otp`, {
    agent_id: agentId
  });
}

export function logDeliveryAction(trackingId: string, actionType: string): Promise<{ success: boolean; message: string }> {
  return apiPost<{ success: boolean; message: string }>(`/staff/action`, { 
    tracking_id: trackingId, 
    action_type: actionType 
  });
}

export function getStaffAnalytics(agentName = "Rohan Sharma"): Promise<StaffAnalytics> {
  return apiGet<StaffAnalytics>(`/staff/analytics`, { agent: agentName });
}

export function getPriorityRoutes(agentLat: number, agentLng: number): Promise<PriorityStop[]> {
  return apiGet<PriorityStop[]>(`/staff/priority-route`, { lat: agentLat, lng: agentLng });
}

export function getStopEta(trackingId: string, agentLat: number, agentLng: number): Promise<EtaResponse> {
  return apiGet<EtaResponse>(`/staff/eta/${trackingId}`, { lat: agentLat, lng: agentLng });
}

export function logVoiceCommand(agentId: string, command: string, success: boolean): Promise<{ success: boolean }> {
  return apiPost<{ success: boolean }>(`/staff/voice-log`, { agent_id: agentId, command, success });
}

export function updateAgentLocation(agentId: string, lat: number, lng: number, speed = 0, batteryLevel = 100): Promise<{ success: boolean }> {
  return apiPost<{ success: boolean }>(`/location/update`, { agent_id: agentId, lat, lng, speed, battery_level: batteryLevel });
}

export function getLiveLocation(agentId: string): Promise<{ lat: number; lng: number; speed: number; battery_level: number }> {
  return apiGet<{ lat: number; lng: number; speed: number; battery_level: number }>(`/location/live/${agentId}`);
}

export function reportIncident(report: IncidentReport): Promise<{ success: boolean; message: string }> {
  return apiPost<{ success: boolean; message: string }>(`/incidents/report`, report);
}

export function getVoiceLogs(agentName = "Rohan Sharma"): Promise<any[]> {
  return apiGet<any[]>(`/staff/voice-logs`, { agent: agentName });
}

// Logic for optimized route remains locally as it performs calculations
export async function getOptimizedRoute(agentName = "Rohan Sharma"): Promise<any> {
  // Let's resolve standard optimized routing sequence
  const stops = await getStaffDeliveries(agentName);
  const active = stops.filter(s => s.status === "current" || s.status === "upcoming");
  
  // Starting point coords (Pune sorting hub)
  let currLat = 18.5204;
  let currLng = 73.8567;
  
  const optimized = [];
  const remaining = [...active];
  let totalDist = 0;
  
  while(remaining.length > 0) {
    let closestIdx = -1;
    let minDist = Infinity;
    
    for(let i = 0; i < remaining.length; i++) {
      const s = remaining[i];
      // Haversine approx
      const R = 6371;
      const dLat = ((s.dest_lat - currLat) * Math.PI) / 180;
      const dLng = ((s.dest_lng - currLng) * Math.PI) / 180;
      const a = Math.sin(dLat/2) * Math.sin(dLat/2) + Math.cos(currLat*Math.PI/180) * Math.cos(s.dest_lat*Math.PI/180) * Math.sin(dLng/2) * Math.sin(dLng/2);
      const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
      const dist = R * c;
      if (dist < minDist) {
        minDist = dist;
        closestIdx = i;
      }
    }
    
    if (closestIdx !== -1) {
      const s = remaining.splice(closestIdx, 1)[0];
      totalDist += minDist;
      optimized.push(s);
      currLat = s.dest_lat;
      currLng = s.dest_lng;
    }
  }
  
  const polyline: number[][] = [[18.5204, 73.8567]];
  optimized.forEach(s => {
    polyline.push([s.dest_lat, s.dest_lng]);
  });
  
  return {
    optimized_order: optimized,
    total_distance_km: Number(totalDist.toFixed(2)),
    duration_text: `${Math.round((totalDist / 50) * 60)} mins`,
    polyline
  };
}